### DISCUSSION TITLE
Handling of dm-verity root hash

### DISCUSSION THREAD

**Discussion by mprt** [rauc/rauc#1185] [Q&A] [answered]
Hi,

This question is only somewhat related to RAUC, but I hope it's not too brash to ask it here...

I've read a couple of times that people are using RAUC to update dm-verity-protected partitions.
I understand that this is a different layer than RAUC's dm-verity protection.

I'm wondering how people handle the root hash when updating.
While you can embed the verity meta-data in the partion, the root hash (and optional signature) have to be stored separately.

- If you update the dm-verity-partition, where/how do you store/modify the root hash?
- If you don't use root hash signatures, then the root hash has to be protected through the secure boot chain, so you necessarily have to change the locatio where the root hash is stored, as well, right? (e.g. FIT-image, kernel, bootloader scripts)
- If you use root hash signatures, where do you store the root hash and the signature?
- (completely off-topic, sorry:) If you put the kernel into the rootfs partition, can you still dm-verity-protect the partition? Do you know if U-Boot/Barebox can handle (or at least read) dm-verity?

Let me know if this is the wrong place for those questions. 
In any case I'm grateful for any answers - my research so far has been surprisingly fruitless...

Matthias

**Comment by jluebbe** [comment, marked as answer]
> I've read a couple of times that people are using RAUC to update dm-verity-protected partitions. I understand that this is a different layer than RAUC's dm-verity protection.

Yes, RAUC is only involved to update the kernel and rootfs slots.

> I'm wondering how people handle the root hash when updating. While you can embed the verity meta-data in the partion, the root hash (and optional signature) have to be stored separately.
> 
> * If you update the dm-verity-partition, where/how do you store/modify the root hash?

Usually in a signed image which contains kernel + initramfs (+ DTB if needed). Without a signed kernel + initramfs, you'd have a gap in your verification chain.

> * If you don't use root hash signatures, then the root hash has to be protected through the secure boot chain, so you necessarily have to change the location where the root hash is stored, as well, right? (e.g. FIT-image, kernel, bootloader scripts)

Yes. In my opinion, you want to bind the kernel binary to a given rootfs anyway to avoid mix-and-match attacks.

> * If you use root hash signatures, where do you store the root hash and the signature?

You could use a separate partition, with the corresponding rootfs as the RAUC parent slot.

> * (completely off-topic, sorry:) If you put the kernel into the rootfs partition, can you still dm-verity-protect the partition? Do you know if U-Boot/Barebox can handle (or at least read) dm-verity?

If you store the root hash somewhere else, that would be possible. Otherwise, you'd have a circular dependency.

As far as I know, neither U-Boot nor Barebox support verifying against the dm-verity hash tree.

In practice, the simpler approach is preferred: Separate partitions for FIT (kernel+initramfs+DTB) and rootfs, always updated together.

> Let me know if this is the wrong place for those questions. In any case I'm grateful for any answers - my research so far has been surprisingly fruitless...

A few years ago, I talked about the basics at ELC (https://www.youtube.com/watch?v=z_y84v9076c). While upstreaming the the signing infrastructure, I made a demo layer with FIT image + dm-verity + rauc: https://github.com/jluebbe/meta-code-signing/tree/master/meta-code-signing-demo. Note that it is a *demo* and will need changes for a real product.
